src/rest/openapi.yaml

```diff
@@ -0,0 +1,455 @@
+          content:
+            application/json:
+              schema:
+                type: number
+                description: Current state
+                example: 4
+  /node/network-name:
+    get:
+      tags:
+        - node
+      summary: Thread network name this node is part of.
+      responses:
+        "200":
+          description: Successful operation
+          content:
+            application/json:
+              schema:
+                type: string
+                description: Thread network name.
+                example: "OpenThread-e445"
+  /node/leader-data:
+    get:
+      tags:
+        - node
+      summary: Gets the network's leader data.
+      responses:
+        "200":
+          description: Successful operation
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/LeaderData"
+  /node/ext-panid:
+    get:
+      tags:
+        - node
+      summary: Extended PAN ID.
+      responses:
+        "200":
+          description: Successful operation
+          content:
+            application/json:
+              schema:
+                type: string
+                description: 8-byte extended PAN ID as hex string.
+                example: "3CAB144450CF407E"
+  /node/num-of-router:
+    get:
+      tags:
+        - node
+      summary: Get number of router devices
+      responses:
+        "200":
+          description: Successful operation
+          content:
+            application/json:
+              schema:
+                type: number
+                description: Number of routers
+                example: 1
+  /node/dataset/active:
+    get:
+      tags:
+        - node
+      summary: Get current active operational dataset
+      responses:
+        "200":
+          description: Returns currently active opertional dataset
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Dataset"
+            text/plain:
+              schema:
+                $ref: "#/components/schemas/TlvDataset"
+        "204":
+          description: No active operational dataset
+    post:
+      tags:
+        - node
+      summary: Create new active opertional dataset
+      description: |-
+        Creates new active operational dataset on the current node. Only allowed if the Thread node is inactive. Create a
+        pending dataset if you would like to change the current active opertional dataset if the Thread node is active.
+
+        Default parameters are choosen for parameters not set in the request body.
+      requestBody:
+        description: |-
+          Operational dataset object that will be stored as active operational dataset. JSON keys which are not set will be initialized with
+          default or random values.
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/Dataset"
```

Comment:
What about supporting "#/components/schemas/DatasetTlv" as well. Thread Active Operational Dataset is typically shared in form of the TLV byte array/HEX string, and it will be useful to commissioning the BR device directly with the dataset TLVs.

Reply:
It can be done with the `PUT` method.

Reply:
But as said in https://github.com/openthread/ot-br-posix/pull/1682#issue-1510058174, POST will create a new dataset but PUT won't. So for commissioning a new BR device, we need to use the POST method and it's more common that a TLV format dataset will be provided for commissioning the BR device. That's why I think TLV format for POST can be useful.

Reply:
@agners What do you think about this?

Reply:
This use case is covered, for the TLV format, we accept `PUT` to create a new active dataset, see:
https://github.com/agners/ot-br-posix/blob/add-dataset-rest-api/src/rest/resource.cpp#L590

Ideally we should only accept `PUT`, as we do not create a sub-item (in a RESTful API `POST` is meant to create a new item underneath the path we are using, this is not the case here). However, for JSON format I need a way to differentiate between creating a new dataset and updating an existing. Since creating a new dataset is not idempotent (since some values are generated random), I feel `POST` is a good choice for that particular use case, even though it doesn't create a sub-item.

Reply:
> Ideally we should only accept PUT, as we do not create a sub-item (in a RESTful API POST is meant to create a new item underneath the path we are using, this is not the case here). 

Make sense.

> However, for JSON format I need a way to differentiate between creating a new dataset and updating an existing.

Why JSON format will be different? I don't think it's a good idea to have any semantic implicities depending on the request payload format. And, a Dataset TLV binary array may include partial dataset as well.

On the other hand, I think the behavior of PUT can be defined as below to make it `idempotent`:
1. create a new random dataset if no existing dataset
2. update the (existing) dataset with the request pyaload

This is `idempotent`, because we are in the same state if the same PUT requests have been processed for multiple times.

Reply:
Hm, merging dataset in TLV format is not something we have code for, from what I can tell. In my mind, when using dataset format one almost always handles complete configuration (e.g. from Android API or somewhere else).

Also, if we only use `PUT`, how can we avoid an update (e.g. if one of the fields should no longer be set/cleared)?

> 1. create a new random dataset if no existing dataset

Google defines Idempotence as no matter how many times you execute them, you achieve the same result. But if random data is generated, that would not be idempotent IMHO (the reason I went for `POST` operation for the dataset JSON API).

Reply:
> Hm, merging dataset in TLV format is not something we have code for, from what I can tell. In my mind, when using dataset format one almost always handles complete configuration (e.g. from Android API or somewhere else).

Almost true, but the RESTful API isn't necessary to make assumptions on Android API.

> Also, if we only use PUT, how can we avoid an update (e.g. if one of the fields should no longer be set/cleared)?

Sorry, I didn't get your point, could you elaborate?

> Google defines Idempotence as "no matter how many times you execute them [operations], you achieve the same result". But if random data is generated, that would not be idempotent IMHO (the reason I went for POST operation for the dataset JSON API).

It's okay, because only the first the PUT request will cause generation of random dataset, the second PUT with the same payload will simply do nothing because we have an existing dataset now.

Reply:
> > Also, if we only use PUT, how can we avoid an update (e.g. if one of the fields should no longer be set/cleared)?
> 
> Sorry, I didn't get your point, could you elaborate?
> 

Currently, when using Dataset TLV, the user controls all the fields which are being set (e.g. if no active operational data set is present, and the user provides a operational dataset without, say, `ExtPanId`, after the operation the active operational dataset will be without `ExtPanId` (this might be a bad example as this is then not a valid configuration? My point is the dataset TLV is applied as is, nothing more, nothing less). It won't be possible to "clear" fields, as we only have the `PUT` method and always use the merge operation.

If we adopt the behavior of `application/json`, it will mean that we generate a dataset and merge the provided one always. The user can not 100% be certain that only the fields of the provided dataset TLV get applied.

Maybe this is the behavior we want, I am just pointing out differences.

> > Google defines Idempotence as "no matter how many times you execute them [operations], you achieve the same result". But if random data is generated, that would not be idempotent IMHO (the reason I went for POST operation for the dataset JSON API).
> 
> It's okay, because only the first the PUT request will cause generation of random dataset, the second PUT with the same payload will simply do nothing because we have an existing dataset now.

Hm, I see, so you'd argue that makes it idempotent. In my mind because randomness is involved, the operation is not idempotent by definition. But I can also live with your interpretation :smile: 

If we are being pedantic, we violate HTTP/REST in some way or another anyways (when using `POST`, because we are not creating a subordinate resource of the resource identified by the Request-URI, when using `PUT` because randomness can be seen as not idempotent operation).

I see your point in making the JSON (`application/json`) and Dataset TLV (`text/plain`) format behave the same way. So following that approach we'd end up with:

- Only `PUT` operation is supported
- If the target dataset doesn't exist, one will get created automatically and the provided fields will be merged (dataset and dataset TLV, `201 Created` is returned)
- If the target dataset exists the provided fields will be merged (dataset and dataset TLV, `200 OK` is returned)
- `PUT` on active dataset is rejected if Thread is running `409 Status Conflict`
- `PUT` on pending dataset is rejected if no delay attribute is provided (dataset and dataset TLV) `400 Bad Request`

Reply:
> Maybe this is the behavior we want, I am just pointing out differences.

Thanks! Yeah, setting partial operational dataset is discouraged. If you want to support this, you should support the JSON format as well.

> Hm, I see, so you'd argue that makes it idempotent. In my mind because randomness is involved, the operation is not idempotent by definition

You are probably referring to "deterministic" rather than "idempotent"?
> Idempotence ([UK](https://en.wikipedia.org/wiki/British_English): [/ˌɪdɛmˈpoʊtəns/](https://en.wikipedia.org/wiki/Help:IPA/English),[[1]](https://en.wikipedia.org/wiki/Idempotence#cite_note-1) [US](https://en.wikipedia.org/wiki/American_English): [/ˈaɪdəm-/](https://en.wikipedia.org/wiki/Help:IPA/English))[[2]](https://en.wikipedia.org/wiki/Idempotence#cite_note-2) is the property of certain [operations](https://en.wikipedia.org/wiki/Operation_(mathematics)) in [mathematics](https://en.wikipedia.org/wiki/Mathematics) and [computer science](https://en.wikipedia.org/wiki/Computer_science) whereby they can be applied multiple times without changing the result beyond the initial application.

This is the wifi definition of `idempotent`, I don't think my PUT operation is violating this :)

Reply:
> - Only PUT operation is supported
> - If the target dataset doesn't exist, one will get created automatically and the provided fields will be merged (dataset and dataset TLV, 201 Created is returned)
> - If the target dataset exists the provided fields will be merged (dataset and dataset TLV, 200 OK is returned)
> - PUT on active dataset is rejected if Thread is running 409 Status Conflict
> - PUT on pending dataset is rejected if no delay attribute is provided (dataset and dataset TLV) 400 Bad Request

All looks good except returning 201 - this code should only be used for response of a POST operation? https://restfulapi.net/http-status-201-created:
> HTTP Status 201 indicates that as a result of HTTP POST request, one or more new resources have been successfully created on the server.

Reply:
SGTM!